pkg/apiserver/clusterinfo/host.go

```diff
@@ -0,0 +1,381 @@
+		return []string{}
+	}
+	return strings.Split(path, string(filepath.Separator))
+}
+
+func inferPartition(dataDir string, diskMap PartitionMap) Partition {
+	var targetDisk Partition
+	var minRelativePathLength = math.MaxInt64
+
+	for _, disk := range diskMap {
+		rel, err := filepath.Rel(disk.Path, dataDir)
+		if err != nil {
+			continue
+		}
+		var relativePathLength int
+		for _, dir := range splitPath(rel) {
+			if dir == ".." {
+				relativePathLength = -1
+				break
+			} else {
+				relativePathLength++
+			}
+		}
+		if relativePathLength == -1 {
+			continue
+		}
+		if relativePathLength < minRelativePathLength {
+			minRelativePathLength = relativePathLength
+			targetDisk = disk
+		}
+	}
+
+	return targetDisk
+}
+
+// HostMap map host ip to all instance on it
+// e.g. "127.0.0.1" => []Instance{...}
+type HostMap map[string][]Instance
+
+func loadHosts(db *gorm.DB) (HostMap, error) {
+	hostMap := make(HostMap)
+	var rows []Instance
+	if err := db.Table("INFORMATION_SCHEMA.CLUSTER_INFO").Find(&rows).Error; err != nil {
+		return nil, err
+	}
+	for _, row := range rows {
+		ip := parseIP(row.Address)
+		instances, ok := hostMap[ip]
+		if !ok {
+			instances = []Instance{}
+		}
+
+		instances = append(instances, Instance{
+			Address:    row.Address,
+			ServerType: row.ServerType,
+		})
+		hostMap[ip] = instances
+	}
+
+	return hostMap, nil
+}
+
+func parseIP(addr string) string {
+	return strings.Split(addr, ":")[0]
+}
+
+// CPUCoreMap map host ip to its cpu logical cores number
+// e.g. "127.0.0.1" => 8
+type CPUCoreMap map[string]int
+
+// Memory map host ip to its Memory detail
+// e.g. "127.0.0.1" => &Memory{}
+type MemoryMap map[string]*Memory
+
+// CPUUsageMap map host ip to its cpu usage
+// e.g. "127.0.0.1" => &CPUUsage{ Idle: 0.1, System: 0.1 }
+type CPUUsageMap map[string]*CPUUsage
+
+type ClusterTableModel struct {
+	Instance   string `gorm:"column:INSTANCE"`
+	DeviceName string `gorm:"column:DEVICE_NAME"`
+	DeviceType string `gorm:"column:DEVICE_TYPE"`
+	Name       string `gorm:"column:NAME"`
+	Value      string `gorm:"column:VALUE"`
+}
+
+const ClusterLoadCondition = "(device_type = 'memory' and device_name = 'virtual') or (device_type = 'cpu' and device_name = 'usage')"
+
+func queryClusterLoad(db *gorm.DB) (MemoryMap, CPUUsageMap, error) {
+	memoryMap := make(MemoryMap)
+	cpuMap := make(CPUUsageMap)
+	var rows []ClusterTableModel
+	if err := db.Table("INFORMATION_SCHEMA.CLUSTER_LOAD").
+		Where(ClusterLoadCondition).Find(&rows).Error; err != nil {
+		return nil, nil, err
+	}
+
+	for _, row := range rows {
+		switch {
+		case row.DeviceType == "memory" && row.DeviceName == "virtual":
+			saveMemory(row, &memoryMap)
+		case row.DeviceType == "cpu" && row.DeviceName == "usage":
+			saveCPUUsageMap(row, &cpuMap)
+		default:
+			continue
+		}
+	}
+	return memoryMap, cpuMap, nil
+}
+
+func saveMemory(row ClusterTableModel, m *MemoryMap) {
+	ip := parseIP(row.Instance)
+
+	memory, ok := (*m)[ip]
+	if !ok {
+		memory = &Memory{}
```

Comment:
A small optimization:

```golang
memory, ok := (*m)[ip]
if !ok {
    memory = &Memory{}
    (*m)[ip] = memory
}   
```

No need to re-assign `(*m)[ip] = memory` when entry exists.